### ISSUE TITLE
question about the scope of mu and sigma,and the order of quantization？

### ISSUE THREAD

**Issue by smileformylove** [JooyoungLeeETRI/CA_Entropy_Model#3]
I was surprised that your work is just the same as google's(David Minnen). What I doubt is whether every symbol(after quantization) need a single mu and sigma.I just want to do like this way,but I think it consumes too much time.When I think over several symbols(maybe one channel is a group) with a single mu and sigma,I doubt the distribution may have deviation?Another question is about  the input of the model of z in your paper?Your input is after quantization,but David has no quantization.Input like this way could improve the accuracy of distribution?So,could you give me some hints?

**Comment by JooyoungLeeETRI** [comment]
I was surprised when I meet Minnen's work, too. :-) With regard to similarity and differences between two works, please refer to OpenReview (https://openreview.net/forum?id=HyxKIiAqYQ). You can find huge amount of discussions there. In short, we agree that the two approaches are very similar, but fortunately, ICLR chairs decided to treat our work as a concurrent work, because two papers have been open to public within one month of time difference.

Regarding your main questions, our models estimates different mu and sigma values for every single component in y_hat. However, for simplicity, our f network estimates all mu and sigma values together for y_hat components which are located at a same spatial point. Note that this does NOT mean y_hat components at a same spatial point share mu and sigma values. For example, when M(the number of channels in y_hat) is 192, f estimates 192 mu values and 192 sigma values together, at one feed-forward process. Please refer to our test code to find this part.

Regarding your second question, the inputs to our entropy model (PMF estimating function) is noisy representations(y_tilde). I guess that's the same as other works. We use the quantized z only for h_s. From our experiments, we found that, when training, rounding inputs(y and z) into the transform functions g_s and h_s improves the compression performance. You can find this in appendix 6.2 of our paper. We think that this removes the mismatches between training and testing.

Please refer to our paper for details: https://arxiv.org/abs/1809.10452
Thank you!!

**Comment by JooyoungLeeETRI** [comment]
One more training skill that we used is that we designate a certain number (16 in the experiments)
of random spatial points as the representatives per training step, as specified in section 4.1 of our paper. For example, you use just 16 or 32 points to train the network. In this case, the training process will be much faster compared to when you calculate the rate term over the entire area of y_tilde (16 by 16 points to calculate in the case of 256 by 256 training inputs)

Hope this helps!

**Comment by smileformylove** [comment]
Thanks so much for your reply,and congratulations on your paper accepted by ICLR2019.Your link to OpenReview helps me a lot,and I have to read it many times.

That's interesting when I use the quantized y for h_a,the quality of reconstructed image is also better w.r.t. y in my experiment.Your reason about  the mismatches between training and testing,but I can not find the relationship between these. Does there exist any mathematics theory in it?

All in all,your training skill is precious to me,and I will try it later!

**Comment by JooyoungLeeETRI** [comment]
Oh, you asked inputs into h_a. That's a bit different from inputs to g_s and h_s. We used quantized y (y_hat) into h_a just because the ultimate goal of our model is to estimate dist. parameters of y_hat, not underlying y. Only the other two quantized inputs(y_hat and z_hat) into g_s and h_s may be related with train/test mismatch.

Frankly, we conducted experiments over different types of inputs to g_s and h_s, but did not over the inputs into h_a. We just configured it by intuition. I'm wondering how much difference your experiments show:-)
 
Anyway, it's really good to hear this thread helps you.
Thank you!

**Comment by smileformylove** [comment]
When I use y_hat for h_a,PSNR has improved by about 2db,and the visualization quality with eyes is better,but MS-SSIM decreases by about 1db with mse optimization. I have not calculated the bpp at that situation.

I reread your answer about mu and sigma with the code,I find I misunderstood the same spatial point.Besides I find another question:-),In equation (5),I think the sigma for z_hat subject to i.i.d,but how does the training process guarantee this?At the same time,I think this simple entropy dependents on a powerful training dataset, It doesn't change with a new picture for testing, is it?

At last,thank you very much.You give me hints though your paper is enough for details!

**Comment by JooyoungLeeETRI** [comment]
In my opinion, 2db of quality difference is too much. From our understanding, and from our experiments on another model we implemented before, quantized inputs to h_a did not noticeably affects the performance. That's why we did not emphasize this point in our paper. 
Regarding p(z), equation (5) is just our assumed prior (hyper prior). Training process does not guarantee that real distributions of z_i are independent from each other. However, from the compression point of view, underlying dependency of real distribution also means "redundancy", so it's the target to be removed by the training process. Therefore, if the network is well-trained, most redundancy will disappear:-)

**Comment by smileformylove** [comment]
Oh,some parameters in my model may effect the results,and I think z_hat is independent from each other in ideal situation.That means "redundancy" have been removed!Thus their distribution is also indenpendent,and the trainable sigma has no relationship with others. That's my understanding.

Anyway,thank you for your answers!
